SWLOR.Game.Server/Feature/AIDefinition/GenericAIDefinition.cs

```diff
@@ -32,485 +34,220 @@ static float CalculateAverageHP(uint creature)
+            {
+                return (usableFeats[randomFeat], target);
+            }
+            else
+            {
+                return (usableFeats[randomFeat], self);
             }
+        }
 
-            // Striking Cobra
-            if (CheckIfCanUseFeat(self, self, FeatType.StrikingCobra3))
-            {
-                return (FeatType.StrikingCobra3, self);
-            }
-            if (CheckIfCanUseFeat(self, self, FeatType.StrikingCobra2))
-            {
-                return (FeatType.StrikingCobra2, self);
-            }
-            if (CheckIfCanUseFeat(self, self, FeatType.StrikingCobra1))
-            {
-                return (FeatType.StrikingCobra1, self);
-            }
+        /// <summary>
+        /// Checks whether a creature can use a specific feat.
+        /// Verifies whether a creature has the feat, meets the condition, and can use the ability.
+        /// </summary>
+        /// <param name="creature">The creature to check</param>
+        /// <param name="target">The target of the feat</param>
+        /// <param name="feat">The feat to check</param>
+        /// <param name="condition">The custom condition to check</param>
+        /// <returns>true if feat can be used, false otherwise</returns>
+        private static bool CheckIfCanUseFeat(uint creature, uint target, FeatType feat, Func<bool> condition = null)
+        {
+            if (!GetHasFeat(feat, creature)) return false;
+            if (condition != null && !condition()) return false;
+            if (!GetIsObjectValid(target)) return false;
 
-            // Slam
-            if (CheckIfCanUseFeat(self, target, FeatType.Slam3))
-            {
-                return (FeatType.Slam3, target);
-            }
-            if (CheckIfCanUseFeat(self, target, FeatType.Slam2))
-            {
-                return (FeatType.Slam2, target);
-            }
-            if (CheckIfCanUseFeat(self, target, FeatType.Slam1))
-            {
-                return (FeatType.Slam1, target);
-            }
+            var targetLocation = GetLocation(target);
+            var abilityDetail = Ability.GetAbilityDetail(feat);
+            var effectiveLevel = Perk.GetEffectivePerkLevel(creature, abilityDetail.EffectiveLevelPerkType);
+            return Ability.CanUseAbility(creature, target, feat, effectiveLevel, targetLocation);
+        }
 
-            // Spinning Whirl
-            if (CheckIfCanUseFeat(self, target, FeatType.SpinningWhirl3))
-            {
-                return (FeatType.SpinningWhirl3, target);
-            }
-            if (CheckIfCanUseFeat(self, target, FeatType.SpinningWhirl2))
-            {
-                return (FeatType.SpinningWhirl2, target);
-            }
-            if (CheckIfCanUseFeat(self, target, FeatType.SpinningWhirl1))
-            {
-                return (FeatType.SpinningWhirl1, target);
-            }
+        private const string CURRENT_WP_VARNAME = "WP_CUR";
```

Comment:
Constants don't need to be capitalized. Can call this CurrentWaypointVariableName or something similar. Even if it's long, I prefer the names to be descriptive and to avoid acronyms in most cases. At a bare minimum remove the all caps please.